Separate configs are generated for each world.

Sharpcastle33 opened the pull request.

Commit:
Added per-world configs. They generate themselves. config.yml is global.

Sharpcastle33 committed:
Cleaned up code and fixed some things for a better merge.

Sharpcastle33 committed:
Trying to fix config problems

Some configs fail to generate, and there are some yml errors in the default-world config.

Sharpcastle33 committed:
Configs now generate, but existing configs aren't being loaded.

Sharpcastle33 committed:
Still trying to figure out this NPE

config.java line 68

ProgrammerDan commented on src/main/java/com/github/devotedmc/hiddenore/tracking/BreakTracking.java:
```diff
@@ -40,54 +30,61 @@ public BreakTracking() {
+		}
+		catch(Exception e) {
+			HiddenOre.getPlugin().getLogger().warning("Error enabling break tracking. Stacktrace:");
+			HiddenOre.getPlugin().getLogger().warning(e.getLocalizedMessage());
```
Generally better here to use the baked in error formatter, like so:

HiddenOre.getPlugin().getLogger().warning("Error enabling break tracking. Stacktrace:", e);

ProgrammerDan commented on src/main/java/com/github/devotedmc/hiddenore/Config.java:
```diff
@@ -47,23 +51,49 @@ private Config() {
+					//TODO pretty sure this whole scope here doesn't work.
+					//TODO double check if this is supposed to be the specific config or default-world.yml for the reader. I'm too tired to be able to follow this around in my head right now.
+					HiddenOre.getPlugin().getLogger().info("Loading configs for world " + world.getName());
+					Reader worldExistsReader = new InputStreamReader(HiddenOre.getPlugin().getResource(String.format("%s-config.yml", world.getName())));
```
getResource() pulls from the .jar file -- not the "file system" iirc.

Just use 

```
Reader worldExistsReader = new InputStreamReader(new FileReader(saveFile));
```

ProgrammerDan commented on src/main/java/com/github/devotedmc/hiddenore/Config.java:
```diff
@@ -47,23 +51,49 @@ private Config() {
+					HiddenOre.getPlugin().getLogger().info("Loaded Configs for world " + world.getName());
+				}else{
+					HiddenOre.getPlugin().getLogger().info("Configs do not exist for world " + world.getName() + ", generating them...");
+					File defaultWorldFile = new File(HiddenOre.getPlugin().getDataFolder(), "default-world.yml");
```
In the same vein, this stuff here kind of does the exact opposite of what you might expect. It checks for the literal file-system existence of the "jar resource" default-world.yml. ... so it's probably not going to work.

Try instead:

```
try (Reader defaultWorldReader = new InputStreamReader(HiddenOre.getPlugin().getResource("default-world.yml"));) {
    configurations.put(world.getName(), YamlConfiguration.loadConfiguration(defaultWorldReader));
    configurations.get(world.getName()).save(new File(HiddenOre.getPlugin().getDataFolder(), String.format("%s-config.yml", world.getName())));
} catch (IOException ioe) {
    HiddenOre.getPlugin().getLogger().warning(String.format("A YAML file could not be created for the world named %s, default-world.yml missing from HiddenOre jar.", world.getName()));
}
```

bonus points: the try-with pattern above will close the resource when done ... I recommend using it in other places where applicable.

ProgrammerDan reviewed (changes requested):
Seems solid, few small changes and I added some notes that might help with the configuration confusion.

Sharpcastle33 committed:
Cleaned up more code for a better merge

Per world configs appear to be affecting those of other worlds than their own -- investigate.

ProgrammerDan commented on src/main/java/com/github/devotedmc/hiddenore/Config.java:
```diff
@@ -65,18 +70,24 @@ public static void loadConfig() {
 					HiddenOre.getPlugin().getLogger().info("Loaded Configs for world " + world.getName());
 				}else{
 					HiddenOre.getPlugin().getLogger().info("Configs do not exist for world " + world.getName() + ", generating them...");
 					File defaultWorldFile = new File(HiddenOre.getPlugin().getDataFolder(), "default-world.yml");
```
Not to beat a horse, but, I'm 60% sure the literal file default-world.yml won't exist on the actual disk-filesystem, but only in the "JAR" filesystem. To access the JAR filesystem, use the .getResource() syntax. Checking the data folder is unlikely to find the default file which will never be "unpacked" there...

ProgrammerDan commented on src/main/java/com/github/devotedmc/hiddenore/tracking/BreakTracking.java:
```diff
@@ -82,8 +82,7 @@ public void load() {
 		catch(Exception e) {
-			HiddenOre.getPlugin().getLogger().warning("Error enabling break tracking. Stacktrace:");
-			HiddenOre.getPlugin().getLogger().warning(e.getLocalizedMessage());
+			HiddenOre.getPlugin().getLogger().warning("Error enabling break tracking. Stacktrace:" + e.getLocalizedMessage());
```
_almost_. Use this exactly:

```HiddenOre.getPlugin().getLogger().warning("Error enabling break tracking.", e);```

and let the logger format the message.

ProgrammerDan reviewed (changes requested):
Very close, sorry for delay on this.

Sharpcastle33 commented:
Don't worry about it man, there are some other problems I need to address. I did some quick cursory testing and it seemed as if all configs were being used for the overworld? I looked around in BMS' code but couldn't figure out how it chose which config to realize into drops. I need to look into it some more but I haven't had time the past two days.

ProgrammerDan commented:
It happens quickly -- the decision is done on point of break, where it picks the configuration to use, iirc. Been a while since I fully reviewed all points of this PR myself, so I could be wrong, but I don't remember it being missing :)

Sharpcastle33 commented:
I'm fairly certain now that no checks for which config to use are made. For one, it looks like it's currently set up so each config *overwrites the last* and that only the last config loaded is used. Not sure how to best quote code here especially because it's a bit messy, but for each world on the server *doLoadWorldConfig()* is called, which explicitly overwrites `instance` with the world's config. Block breaks appear to also only refer to `instance`.

I'm going to setup a system to store multiple configs and then decide on break which config to refer to. Hopefully if I don't break anything it should be good to go.

Sharpcastle33 committed:
Began to implement the ability to maintain multiple Config instances

Also fixed global config options

Quick commit so I can backtrack later if I fuck things up

for some reason Map configurations is always empty and Map instances only ever contains the last map loaded (firedim), working on it

Sharpcastle33 committed:
HiddenOre now correctly manages configs for each world.

isDropBlock(...) now correctly returns a config based on the world chosen.

Map instances now works correctly.

I still have a lot of debug messages to remove and I have not confirmed if all configs generate as they are supposed to in the first place.

ProgrammerDan commented:
A few things still seem a little off but I'm too tired ATM to pick them apart. Have you tested? In spite of appearance is it working? What have you tried to verify it works? Thanks!

Sharpcastle33 commented:
I still need to clean up a lot of stuff as well as make sure everything is working -- I made quite a mess with debug messages and such trying to figure out a few kinks which took way more hours than it should've.

I only tested very briefly; I went into the configs and quickly made redstone common in one world, emeralds common in another, and lapis common in a third, and got the correct ore in each world. HiddenOre now maintains a list of configs, and based on the world a block was broken in, hands off a corresponding config to the rest of the plugin. BMS had disabled a lot of the chat prefix functionality and the like, I simply added it as a 'global' config, which made chat messages re-appear, but I haven't thoroughly tested it or anything.

I had one problem that I still honestly don't understand, that took like 2-3 hours to solve. I had been maintaining the list of instances in the Config class because originally the Config instance was maintained there. However, that caused problems when loading individual world configs to a map. Looping through each world and calling ``instances.put(worldName,i)`` (with i being the Config created for the world in question) resulted in me receiving a hashmap with only one world, rather than the several that I expected. This really through me off and took a while to narrow down the cause. I think it probably had to do with the way Java handles scopes or something but I'm honestly not sure. I moved ``instances`` to the HiddenOre class and it worked without changing anything else. Maintaining instances of Configs inside the Config class felt really messy anyways so it was probably for the better.

Oh, I also temporarily disabled the whole clear_ores functionality, because it was called before some of the stuff it relies on is actually registered. I have to look into how I want to best fix that.

ProgrammerDan commented:
Alright sounds good, lmk if you run into roadblocks otherwise, I'll hold further comments for a bit while you sort out the issues you've already identified.

ProgrammerDan commented:
is this still something you intend to do?

Sharpcastle33 commented:
My fork has all the features I need on it, I don't plan on re-implementing clear ores functionality.

ProgrammerDan commented:
I see; so you have no intent to push your changes upstream?

ProgrammerDan commented:
I've spent some time considering these changes. I don't prefer the multiple file approach as defined here; it's a perfectly valid approach in general, but a more elegant and minimal-change approach is possible just injecting optional-world defines into the main config. I've got some ideas sketched out in an uncommitted branch, which will likely go online tonight / tomorrow, including example config. Once I do that, I'll probably close this / these. 

That said, thanks for your efforts on this, and for illustrating well the need for this sort of support!

ProgrammerDan commented:
See latest master commits, v. 1.4.2 release forthcoming. Then, on to 1.13
